packages/compass-workspaces/src/stores/workspaces.ts

```diff
@@ -973,6 +992,105 @@ type DatabaseRemovedAction = {
   namespace: string;
 };
 
+/**
+ * Converts saved workspace state data back to OpenWorkspaceOptions format
+ * for initializing the store
+ */
+function convertSavedStateToOpenWorkspaceOptions(
+  savedState: WorkspacesStateData
+): OpenWorkspaceOptions[] {
+  return savedState.tabs.map((tab) => {
+    const baseTab: Record<string, unknown> = { type: tab.type };
+
+    // Add connection-related fields
+    if (tab.connectionId) {
+      baseTab.connectionId = tab.connectionId;
+    }
+    if (tab.namespace) {
+      baseTab.namespace = tab.namespace;
+    }
+
+    // Add optional fields based on workspace type
+    if (tab.initialQuery) {
+      baseTab.initialQuery = tab.initialQuery;
+    }
+    if (tab.initialAggregation) {
+      baseTab.initialAggregation = tab.initialAggregation;
+    }
+    if (tab.initialPipeline) {
+      baseTab.initialPipeline = tab.initialPipeline;
+    }
+    if (tab.initialPipelineText) {
+      baseTab.initialPipelineText = tab.initialPipelineText;
+    }
+    if (tab.editViewName) {
+      baseTab.editViewName = tab.editViewName;
+    }
+    if (tab.initialEvaluate) {
+      baseTab.initialEvaluate = tab.initialEvaluate;
+    }
+    if (tab.initialInput) {
+      baseTab.initialInput = tab.initialInput;
+    }
+    if (tab.subTab) {
+      baseTab.initialSubtab = tab.subTab;
+    }
+
+    return baseTab as OpenWorkspaceOptions;
+  });
+}
+
+export const loadSavedWorkspaces = (): WorkspacesThunkAction<
+  Promise<void>,
+  RestoreWorkspacesAction
+> => {
+  return async (dispatch, getState, { connections, userData }) => {
+    const savedState = await userData.readOne('saved-workspaces', {
+      ignoreErrors: true,
+    });
+    if (savedState) {
+      const confirm = await showConfirmation({
+        title: 'Reopen closed tabs?',
+        description:
+          'Your connection and tabs were closed, this action will reopen your previous session',
+        buttonText: 'Reopen tabs',
+      });
+
+      const workspacesToRestore: OpenWorkspaceOptions[] = [];
+      const connectionsToRestore: Map<string, ConnectionInfo> = new Map();
+      (confirm
+        ? convertSavedStateToOpenWorkspaceOptions(savedState)
+        : []
+      ).forEach((workspace) => {
+        // If the workspace is tied to a connection, check if the connection exists
+        // and add it to the list of connections to restore if so.
+        if ('connectionId' in workspace) {
+          const connectionInfo = connections.getConnectionById(
+            workspace.connectionId
+          )?.info;
+
+          if (!connectionInfo) {
+            return;
+          }
+
+          connectionsToRestore.set(workspace.connectionId, connectionInfo);
```

Comment:
Super nit: feels like you can just call connect right here, no reason to collect those first as far as I can see

```suggestion
          void connections.connect(connectionInfo);
```

Reply:
Ah okay, I wasn't sure if the connect method deals with duplicates.

Reply:
It does, but the more I think about it, the more I'm on the fence 😆 Maybe keep your version, doesn't hurt to be more explicit here